### DISCUSSION TITLE
Unable to connect file storage to GCS Bucket

### DISCUSSION THREAD

**Discussion by SlitiBrahim** [pocketbase/pocketbase#3022] [Q&A]
Hey!

I am trying to setup the file storage to a Google Bucket but I am encountering an issue where I get the message "Failed to establish S3 connection".

As the endpoint I've set "storage.googleapis.com", I've also set the bucket name and region according to my bucket config.
For the access key field, I've set the email address of the service account I have that has the permissions to manage the bucket objects. For the secret I've set the private key of my service account.

Is that correct?

I've also run the server with `--debug` but I just see the sqlite UPDATE log with the values I've provided.

Please find below the aws go package versions if that can help:
```
github.com/aws/aws-sdk-go v1.44.298 // indirect
github.com/aws/aws-sdk-go-v2 v1.18.1 // indirect
github.com/aws/aws-sdk-go-v2/aws/protocol/eventstream v1.4.10 // indirect
github.com/aws/aws-sdk-go-v2/config v1.18.27 // indirect
github.com/aws/aws-sdk-go-v2/credentials v1.13.26 // indirect
github.com/aws/aws-sdk-go-v2/feature/ec2/imds v1.13.4 // indirect
github.com/aws/aws-sdk-go-v2/feature/s3/manager v1.11.71 // indirect
github.com/aws/aws-sdk-go-v2/internal/configsources v1.1.34 // indirect
github.com/aws/aws-sdk-go-v2/internal/endpoints/v2 v2.4.28 // indirect
github.com/aws/aws-sdk-go-v2/internal/ini v1.3.35 // indirect
github.com/aws/aws-sdk-go-v2/internal/v4a v1.0.26 // indirect
github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding v1.9.11 // indirect
github.com/aws/aws-sdk-go-v2/service/internal/checksum v1.1.29 // indirect
github.com/aws/aws-sdk-go-v2/service/internal/presigned-url v1.9.28 // indirect
github.com/aws/aws-sdk-go-v2/service/internal/s3shared v1.14.3 // indirect
github.com/aws/aws-sdk-go-v2/service/s3 v1.36.0 // indirect
github.com/aws/aws-sdk-go-v2/service/sso v1.12.12 // indirect
github.com/aws/aws-sdk-go-v2/service/ssooidc v1.14.12 // indirect
github.com/aws/aws-sdk-go-v2/service/sts v1.19.2 // indirect
github.com/aws/smithy-go v1.13.5 // indirect
```

It would be amazing if you could help me! Thank you in advance :)

**Comment by ganigeorgiev** [comment]
Unfortunately I can't help with the GCS cloud credentials setup as I haven't personally tried it.

To help you with the debugging you can try to hover on the error in the Admin UI label and check if there is a more detailed raw error message. Something like
![error_tooltip](https://i.imgur.com/MOwyHt9.png)

`--debug` will print the S3 error only when there is a running S3 operation (eg. when checking the S3 connection status in the Admin UI).

You can also check the user suggestion from https://github.com/pocketbase/pocketbase/discussions/1596.

**Reply by SlitiBrahim** [reply]
Thank you for the prompt answer it's really pleasant.

When I hover on the error I get the following message:
<img width="557" alt="Screenshot 2023-07-30 at 17 39 41" src="https://github.com/pocketbase/pocketbase/assets/24780841/035d8af4-fe86-4ce7-95c0-2ace89dd57ac">

Thank you for the resources, I think I will just create my bucket on AWS.

I got 2 more questions please:
- Do you think it's possible to completely disable the file URLs? Like `http://127.0.0.1:8090/api/files/COLLECTION_ID_OR_NAME/RECORD_ID/FILENAME` I don't want the files to be accessible, I know there is this `protected` files feature but I want to completely disallow users to request the files.
- Plus, I would like to store the uploaded files to a bucket and once uploaded, set the URL of the file from the bucket to the pocketbase Db, how can I achieve this?

Thank yo very much

**Reply by ganigeorgiev** [reply]
> Do you think it's possible to completely disable the file URLs? I don't want the files to be accessible, I know there is this protected files feature but I want to completely disallow users to request the files.

If you don't want to use the [Protected files](https://pocketbase.io/docs/files-handling/#protected-files) feature, then you can try to return an error in the `OnFileDownloadRequest` hook, something like:

```go
// go:
app.OnFileDownloadRequest().Add(func(e *core.FileDownloadEvent) error {
	// disallow downloading the file if not admin
	// (note that the preview in the Admin UI may not work because the Authorization: TOKEN header is not sent with the img tag request)
	if e.HttpContext.Get(apis.ContextAdminKey) == nil {
		return apis.NewForbiddenError("You are not allowed to access this resource.", nil)
	}

	return nil
})
```

Or with JS if you use PocketBase v0.17+
```js
onFileDownloadRequest((e) => {
    // disallow downloading the file if not admin
    // (note that the preview in the Admin UI may not work because the Authorization: TOKEN header is not sent with the img tag request)
    if (!e.httpContext.get("admin")) {
        throw new ForbiddenError("You are not allowed to access this resource.")
    }
})
```

---

> Plus, I would like to store the uploaded files to a bucket and once uploaded, set the URL of the file from the bucket to the pocketbase Db, how can I achieve this?

I'm not entirely sure that I understand what do you mean here but general filesystem hooks are planned for the near future (see https://github.com/pocketbase/pocketbase/issues/2387) which would allow performing certain operations after upload/delete.
Please note that the uploaded object prefix is always in the format `{collectionId}/{recordId}/{filename}` so you can construct the bucket address on your own as long as you have the 3 components (which usually are available in the json record data).

**Reply by ganigeorgiev** [reply]
**Edit about the file error check:** Keep in mind that if you return an error via the file hook the file previews in the Admin UI may not work because the `Authorization: TOKEN` header is not sent with the img tag request. My recommendation would be to not return the above error, unless really needed, but to mark your file fields as "Protected" to allow previewing them in the Admin UI.

**Reply by SlitiBrahim** [reply]
Ok I see thank your for the detailed answers.

I am testing the file feature now with the s3 plug:
- Is it normal that when I delete a record that contains a file (on the UI for now) the file is not being deleted from the s3? How can I achieve this?
- I need to save the uploaded file under a specific name I would like to control in the go server, let's say a user uploads a file name "image.png" I would like to persist this file in s3 with a really specific name let's say "12345.png" how can I do this?
- Also, I need to persist the final s3 URL of the file being uploaded in the record containing the file, you said I can construct the URL with post work but I want to do it at time of saving the file to avoid future URL calculation operations, let's say the user uploads his "image.png" file, I want to save it under "12345.png" and I want to set the `fileBucketURL` of my record to `https://s3.amazonaws.com/my-bucket.com/{collection ID}/{recordId}/12345.png`.

Thank you!

**Reply by ganigeorgiev** [reply]
1. No, it is not normal. If you delete the record or the individual file, the file from the S3 should be also deleted. If you are not experiencing this, then it is either a bug (I'll test it sometime later today[0]) OR make sure to check whether your S3 provider has versioning, batched deletes or "deleted files policy" enabled (backblaze for example keep deleted files for 30days - https://www.backblaze.com/cloud-backup/features/extended-version-history).


2. This is not supported at the moment.

3. You can attach to the [OnRecordAfterCreateRequest hook](https://pocketbase.io/docs/go-event-hooks/#onrecordaftercreaterequest), iterate over the file(s) of the record and save the result in some text field. But I'd advice against this. It would be easier to create a helper client-side function that accept a record object and return the url you need, for example:
```js
// eg. getS3Url(record, "test.png")
function getS3Url(record, filename) {
    return `https://s3.amazonaws.com/my-bucket.com/${record.collectionId}/${record.id}/${filename}`
}
```

---

[0]: _I've tested it locally with Wasabi S3 and the deletion is working fine for me. We also have [integration tests](https://github.com/pocketbase/pocketbase/blob/master/apis/record_crud_test.go#L741) that checks for deleted files when deleting record, so I suspect in your case is some retention policy that is enabled by default._

**Comment by SlitiBrahim** [comment]
Thank you!

1. I've double checked and I don't have any versioning:
![image](https://github.com/pocketbase/pocketbase/assets/24780841/24e768be-007f-4ba5-83ed-d698571094ad)

2. Ok

3. Good thankss

**Reply by ganigeorgiev** [reply]
For 1), as mentioned in the edit in https://github.com/pocketbase/pocketbase/discussions/3022#discussioncomment-6588593, I'm not able to reproduce it with Wasabi S3 and locally. We also have tests that ensures the correct behavior. The only case I can think of where a file may not be deleted is if you delete the collection file field (_this is a very low priority and tracked in https://github.com/pocketbase/pocketbase/issues/305_).

You can start PocketBase with `--debug` and check if there is any error message in the console when deleting the record with the file.

**Comment by SlitiBrahim** [comment]
Thanks a lot
